File: 04/04.pl
:- use_module(library(lists)).

range(X, X, [X]).
range(X, Y, [X|Xs]) :- Z is X+1, range(Z, Y, Xs).

summaplist(F, List, Sum) :- maplist(F, List, Results), sumlist(Results, Sum).

full_overlap([List1, List2], 1) :- intersection(List1, List2, List1).
full_overlap([List1, List2], 1) :- intersection(List1, List2, List2).
full_overlap(_, 0).

partial_overlap([List1, List2], 1) :- intersection(List1, List2, [_|_]).
partial_overlap(_, 0).

parse_line(String, [Assignment1, Assignment2]) :-
    split_string(String, ",", "", Parts),
    maplist(parse_assignment, Parts, [Assignment1, Assignment2]).

parse_assignment(String, Range) :-
    split_string(String, "-", "", Parts),
    maplist(atom_number, Parts, [Low, High]),
    range(Low, High, Range).

read_input(Lines) :-
    open("input_full.txt", read, Input),
    read_file_to_list(Input, StringLines),
    maplist(parse_line, StringLines, Lines).

main :-
    read_input(Lines),
    summaplist(full_overlap, Lines, Part1Total),
    print(Part1Total), nl,
    summaplist(partial_overlap, Lines, Part2Total),
    print(Part2Total), nl,
    halt.

read_file_to_list(Stream, []) :- at_end_of_stream(Stream).
read_file_to_list(Stream, [Line|Lines]) :-
    read_line_to_string(Stream, Line),
    read_file_to_list(Stream, Lines).

File: 04/04.hs
import Data.List

parse :: String -> [([Int], [Int])]
parse = map parseLine . lines

parseLine :: String -> ([Int], [Int])
parseLine = partitionWith (uncurry enumFromTo . partitionWith read '-') ','

partitionWith :: (String -> a) -> Char -> String -> (a, a)
partitionWith f d text = let (l, _:r) = break (== d) text in (f l, f r)

fullOverlap, partialOverlap :: [Int] -> [Int] -> Bool
fullOverlap r1 r2 = let i = r1 `intersect` r2 in i == r1 || i == r2
partialOverlap r1 r2 = not . null $ r1 `intersect` r2

main :: IO ()
main = do
  input <- parse <$> readFile "input_full.txt"
  print . length . filter (uncurry fullOverlap) $ input
  print . length . filter (uncurry partialOverlap) $ input
